modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java

```diff
@@ -141,7 +146,12 @@ private void close(Exception cause) {
     /** {@inheritDoc} */
     @Override
     public void onMessage(ByteBuf buf) {
-        processNextMessage(buf);
+        try {
+            processNextMessage(buf);
+        } catch (Throwable t) {
+            buf.release();
+            throw t;
+        }
```

Comment:
`finally`?

Reply:
Sorry, finally will not work this way here. Ignore

Reply:
When everything goes well, the buffer is released by the `try` block on line 247 after decoding the message. The decoding happens later on `asyncContinuationExecutor`, not on a current thread, so we can't use `finally` here.

But when there is an exception (unexpected message type, unexpected response id, etc) we should release the buffer here to avoid the leak, because full message decoding won't happen.

Reply:
Yeah, I undestand it now. Thanks